lib/synapse/service_watcher/zookeeper_poll.rb

```diff
@@ -0,0 +1,81 @@
+require 'synapse/service_watcher/base'
+require 'synapse/service_watcher/zookeeper'
+
+require 'zk'
+require 'thread'
+require 'timeout'
+
+class Synapse::ServiceWatcher
+  class ZookeeperPollWatcher < ZookeeperWatcher
+    def start
+      log.info 'synapse: ZookeeperPollWatcher starting'
+
+      @thread = nil
+      @mutex = Mutex.new
+      @mutex.lock
+
+      zk_connect do
+        @thread = Thread.new {
+          log.info 'synapse: zookeeper polling thread started'
+
+          should_exit = false
+          until should_exit
+            discover
+
+            sleep_duration = @discovery['polling_interval_sec']
+            log.info "synapse: zookeeper polling thread sleeping for #{sleep_duration} seconds"
+
+            # Awake either when
+            # (1) Mutex is released, which occurs in #stop. This means we
+            #     should exit the thread, which will occur on the next
+            #     iteration.
+            # (2) sleep_duration seconds has passed.
+            begin
+              Timeout::timeout(sleep_duration) {
```

Comment:
Needing to release lock to make next polling available, right? If so, should here be 
```
{
  @mutex.unlock
   @should_exit = false
}
```

Reply:
actually, do you think `sleep @sleep_duration` is enough? You don't need `lock` but could only use `@should_exit` to control the liveness of the thread.